### PR TITLE
qt: Migrate to Qt6.

### PR DESCRIPTION
**Pull request by Steveice10**
Migrates the codebase to Qt 6. Most of the changes are pretty minor to move away from deprecated or removed APIs.  The main changes are to camera support, which required some significant revisions for new media device APIs. Some pieces related to pixel formats and frame rates were removed as I couldn't find an equivalent way to manage them in Qt 6, but it shouldn't matter as the actual format conversions and frame timings are already handled by core code.

Any testing is appreciated. A general pass over UI functionality should suffice as this doesn't touch any code outside the Qt frontend. Mainly camera support needs testing across platforms as it changed the most.

Depends on: https://github.com/citra-emu/citra/pull/6435

- - -
This change is [Reviewable](https://reviewable.io/reviews/citra-emu/citra/6418)

### CODE REVIEW

**Comment by ImperatorStorm** [comment]
Built on Arch Linux, very quick testing showed:
UI works fine.
Camera appears to have completely broken, showing only a still black image.

**Comment by Steveice10** [comment]
> Built on Arch Linux, very quick testing showed: UI works fine. Camera appears to have completely broken, showing only a still black image.

Do you know if the camera device is still showing up in the settings or if it's black because it can't find a camera? Helps to know for root causing the issue.

**Comment by ImperatorStorm** [comment]
> > Built on Arch Linux, very quick testing showed: UI works fine. Camera appears to have completely broken, showing only a still black image.
> 
> Do you know if the camera device is still showing up in the settings or if it's black because it can't find a camera? Helps to know for root causing the issue.

Camera shows up in settings and is selectable, though the camera stays black regardless.

**Comment by GPUCode** [comment]
Do local windows builds work after removing the bundled qt binaries?

**Comment by Steveice10** [comment]
> Do local windows builds work after removing the bundled qt binaries?

Probably not unless you install Qt... not sure if we'd want to continue manually rebuilding the externals just for local Windows builds or just direct people to install Qt.

File: src/citra_qt/game_list_p.h [hunk 1]
```diff
@@ -67,7 +67,7 @@ static QPixmap GetDefaultIcon(bool large) {
  */
 static QString GetQStringShortTitleFromSMDH(const Loader::SMDH& smdh,
                                             Loader::SMDH::TitleLanguage language) {
-    return QString::fromUtf16(smdh.GetShortTitle(language).data());
+    return QString::fromUtf16(reinterpret_cast<char16_t*>(smdh.GetShortTitle(language).data()));
```
Comment (lioncash) [inline src/citra_qt/game_list_p.h:70]: As far as I know, there shouldn't be any problem changing GetShortTitle() and GetLongTitle()'s return type to be an array of char16_t (and also converting the struct member over to that as well), which would make these casts unnecessary

**Comment by GPUCode** [comment]
> > Do local windows builds work after removing the bundled qt binaries?
> 
> Probably not unless you install Qt... not sure if we'd want to continue manually rebuilding the externals just for local Windows builds or just direct people to install Qt.
> 
> Under the hood the GitHub action uses this, which seems to be a pretty easy way to setup Qt on most systems: https://github.com/miurahr/aqtinstall

I don't know if requiring an external Qt installation is the best way forward as building citra locally will become a more involved process. On the other hand if this simplifies Qt handling I don't mind it either, using the action script appears to have advantages as well (it's very easy to upgrade the Qt version). I feel this warrants more discussion though.

**Comment by SachinVin** [comment]
I definitely think we should fetch qt from our repo, downloading qt from the website requires you to create an account and what not.

**Comment by Steveice10** [comment]
> I definitely think we should fetch qt from our repo (for local builds), downloading qt from the website requires you to create an account and what not.

With the installer I linked you don’t need to do this.

**Comment by Steveice10** [comment]
As discussed on Discord, have added back the support for automatic Qt download using aqtinstaller, meaning we're on par with previous functionality without having to bundle Qt ourselves anymore.

**Comment by ImperatorStorm** [comment]
Testing on Linux, Citra downloads qt binaries irrespective of `CITRA_USE_BUNDLED_QT` being set to `OFF`.

**Comment by Steveice10** [comment]
> Testing on Linux, Citra downloads qt binaries irrespective of `CITRA_USE_BUNDLED_QT` being set to `OFF`.

Should be fixed

**Comment by ImperatorStorm** [comment]
> > Testing on Linux, Citra downloads qt binaries irrespective of `CITRA_USE_BUNDLED_QT` being set to `OFF`.
> 
> Should be fixed

Fixed, thanks!

**Comment by Steveice10** [comment]
Build images have been updated and CI is all passing, ready for review. Note that the first two commits of this are https://github.com/citra-emu/citra/pull/6435, which this PR depends on.

File: src/citra_qt/game_list.cpp [hunk 2]
```diff
@@ -418,10 +419,10 @@ void GameList::DonePopulating(const QStringList& watch_list) {
     // Workaround: Add the watch paths in chunks to allow the gui to refresh
     // This prevents the UI from stalling when a large number of watch paths are added
     // Also artificially caps the watcher to a certain number of directories
-    constexpr int LIMIT_WATCH_DIRECTORIES = 5000;
+    constexpr qsizetype LIMIT_WATCH_DIRECTORIES = 5000;
     constexpr int SLICE_SIZE = 25;
-    int len = std::min(watch_list.length(), LIMIT_WATCH_DIRECTORIES);
-    for (int i = 0; i < len; i += SLICE_SIZE) {
+    qsizetype len = std::min(watch_list.length(), LIMIT_WATCH_DIRECTORIES);
```
Comment (GPUCode) [inline src/citra_qt/game_list.cpp:424]:
```suggestion
    const qsizetype len = std::min(watch_list.length(), LIMIT_WATCH_DIRECTORIES);
```

File: src/citra_qt/hotkeys.cpp [hunk 3]
```diff
@@ -26,7 +26,7 @@ void HotkeyRegistry::SaveHotkeys() {
 void HotkeyRegistry::LoadHotkeys() {
     // Make sure NOT to use a reference here because it would become invalid once we call
     // beginGroup()
-    for (auto shortcut : UISettings::values.shortcuts) {
+    for (const auto& shortcut : UISettings::values.shortcuts) {
```
Comment (GPUCode) [inline src/citra_qt/hotkeys.cpp:29]: It says above not to use a reference here

**Review by GPUCode — APPROVED** [review]
LGTM

**Comment by Steveice10** [comment]
Rebased

**Comment by GPUCode** [comment]
Conflicts with https://github.com/citra-emu/citra/pull/6435 so utagging

**Comment by Steveice10** [comment]
> Conflicts with #6435 so untagging

Feel free to also just tag this and untag that if you want, this contains the commits from that other PR since it's dependent.

**Comment by GPUCode** [comment]
Okay I assumed we might want to merge that one first since it's smaller, but that's good too

**Comment by Daisouji** [comment]
scaling is off compared to before
Capture

**Comment by Steveice10** [comment]
> scaling is off compared to before

Please provide a screenshot of what it looked like before as well.

**Comment by Daisouji** [comment]
Capture (1)

**Comment by Steveice10** [comment]
This looks like expected behavior to me, Qt 6 is respecting your Windows scaling settings.

**Comment by Daisouji** [comment]
> This looks like expected behavior to me, Qt 6 is respecting your Windows scaling settings.

But yuzu is QT6 and it scales properly.

**Comment by Daisouji** [comment]
> This looks like expected behavior to me, Qt 6 is respecting your Windows scaling settings.

The 2nd pic I posted was the original QT on Nightly, the first is the latest Canary.

**Comment by Steveice10** [comment]
It looks like this commit broke the current workaround by resetting the scale rounding policy when our screen testing QApplication is destructed: https://github.com/qt/qtbase/commit/e1c8ad359d6fdb536732506f18a3ce864db632e4

The screen info also seems to return pre-scaled values now, resulting in the calculated real ratio being 1.

Should be fixed now, I just moved the call to set the policy to after the destructor is called and updated how we calculate the real values.

**Comment by Daisouji** [comment]
> It looks like this commit broke the current workaround by resetting the scale rounding policy when our screen testing QApplication is destructed: [qt/qtbase@e1c8ad3](https://github.com/qt/qtbase/commit/e1c8ad359d6fdb536732506f18a3ce864db632e4)
> 
> The screen info also seems to return pre-scaled values now, resulting in the calculated real ratio being 1.
> 
> Should be fixed now, I just moved the call to set the policy to after the destructor is called and updated how we calculate the real values.

Judging by the artifact, it's much more useable. #6441 also covers scaling.

**Pull request merged**